engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/skipInclude/SkipIncludeTransform.kt

```diff
@@ -0,0 +1,125 @@
+package graphql.nadel.enginekt.transform.skipInclude
+
+import graphql.execution.MergedField
+import graphql.introspection.Introspection
+import graphql.language.Field
+import graphql.nadel.Service
+import graphql.nadel.ServiceExecutionResult
+import graphql.nadel.enginekt.NadelExecutionContext
+import graphql.nadel.enginekt.blueprint.NadelOverallExecutionBlueprint
+import graphql.nadel.enginekt.transform.NadelTransform
+import graphql.nadel.enginekt.transform.NadelTransformFieldResult
+import graphql.nadel.enginekt.transform.artificial.NadelAliasHelper
+import graphql.nadel.enginekt.transform.query.NadelQueryTransformer
+import graphql.nadel.enginekt.transform.result.NadelResultInstruction
+import graphql.nadel.enginekt.transform.skipInclude.SkipIncludeTransform.State
+import graphql.nadel.enginekt.util.resolveObjectTypes
+import graphql.nadel.enginekt.util.toBuilder
+import graphql.normalized.ExecutableNormalizedField
+import graphql.normalized.ExecutableNormalizedField.newNormalizedField
+import graphql.schema.GraphQLSchema
+
+internal class SkipIncludeTransform : NadelTransform<State> {
+    companion object {
+        private const val skipFieldName = "__skip"
+    }
+
+    class State(
+        val aliasHelper: NadelAliasHelper,
+    )
+
+    override suspend fun isApplicable(
+        executionContext: NadelExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        services: Map<String, Service>,
+        service: Service,
+        overallField: ExecutableNormalizedField,
+    ): State? {
+        // This hacks together a child that will pass through here
+        if (overallField.children.isEmpty()) {
+            val mergedField = executionContext.query.getMergedField(overallField)
+            if (hasAnyChildren(mergedField)) {
+                // Adds a field so we can transform it
+                overallField.children.add(createSkipField(executionBlueprint.schema, overallField))
```

Comment:
Ok so Nadel transforms right now aren't designed to transform the children to add artificial fields.

It's not impossible but it's a bit of a mess right now with hydrations etc.

So what this does here is it hacks back in the skipped field that was removed by the normalised query factory. The query traversal will then immediately pick is up and pass it back into `isApplicable`

The new field is added with a specific name that we check, then we transform that field to automatically remove it.

Reply:
So we're more or less adding the field as a child of itself so it will be picked up by a following step of the traverser?
Couldn't we just return a non-null `State` at this point? 🤔

Reply:
Nah unfortunately not. The issue is is with hydrations. If we return a non-null `State` here, then we're targetting the _parent_ field e.g.

```
{
  root {
    thing {
      child @skip(if: true)
    }
  }
}
```

The issue is hydration. Say `thing` is the hydration field, then it kinda just deletes its portion of the tree and no transforms get applied. When it deletes the tree, it executes it again as the children of the hydration and does the entire process from the top down i.e. goes back to `NextgenEngine`.

BUT since this transformation runs _before_ hydration, we will insert the `__typename` when looking at `thing`, which hydration then deletes and moves to a completely new execution. The new execution doesn't have context that the `__typename` is an artificial field.

Anyway sorry if that's a horrible explanation. Kinda complex issue with the way transforms work.

Reply:
Thanks for explaining it Franklin